docs/cudf/source/internals.md

```diff
@@ -1,29 +1,132 @@
 cuDF internals
 ==============
 
-ColumnAccessor
---------------
+The cuDF API closely matches that of the [Pandas](https://pandas.pydata.org/) library.
+Thus, we have the types `cudf.Series`, `cudf.DataFrame` and `cudf.Index` which look and
+feel very much like their Pandas counterparts.
+
+Under the hood, however, cuDF uses data structures very different from Pandas. In this document,
+we describe these internal data structures.
+
+## Column
+
+Columns are cuDF's core data structure and they are modeled after
+the [Apache Arrow Columnar Format](https://arrow.apache.org/docs/format/Columnar.html).
+
+A column represents a sequence of values, any number of which may be "null". Columns are
+specialized based on the type of data they contain. Thus we have `NumericalColumn`, `StringColumn`,
+`DatetimeColumn`, etc.,
+
+A column is composed of the following:
+
+* A **data type**, specifying the type of each element.
+* A **data buffer** that may store the data for the column elements.
+  Some column types do not have a data buffer, instead storing data in the children columns.
+* A **mask buffer** whose bits represent the validity (null or not null) of each element.
+  Columns whose elements are all "valid" may not have a mask buffer. Mask buffers are padded
+  to 64 bytes.
+* A tuple of **children** columns, which enable the representation complex types such as
+  columns with non-fixed width elements (e.g., strings, lists).
+* A *size* indicating the number of elements in the Column.
+* An integer **offset**: a column may represent a "slice" of another column,
+  in which case this offset represents the first element of the slice. The size of
+  the column then gives the extent of the slice. A column that is not a slice
+  has an offset of 0.
+
+For example, the `NumericalColumn` backing a Series with 1000 elements of type 'int32'
+and containing nulls is composed of:
+
+1. A data buffer of size 4000 bytes (sizeof(int32) * 1000)
+2. A mask buffer of size 128 bytes (1000/8 padded to a multiple of 64 bytes)
+3. No children columns
+
+As another example, the `StringColumn` backing the Series
+`['do', 'you', 'have', 'any', 'cheese?']` is composed of:
+
+1. No data buffer
+2. No mask buffer as there are no nulls in the Series
+3. Two children columns:
+   - A column of 8-bit characters `['d', 'o', 'y', 'o', 'u', h' ... '?']`
+   - A column of "offsets" to the characters column (in this case, `[0, 2, 5, 9, 12, 19]`)
+
+## Buffer
+
+The data and mask buffers of a Column represent data in GPU memory (a.k.a *device memory*),
+and are object of type `cudf.core.buffer.Buffer`.
+
+Buffers can be constructed from array-like objects that live either on the host (e.g., numpy arrays)
+or the device (e.g., cupy arrays).
+
+When constructing a Buffer from a host object such as a numpy array, new device memory is allocated:
+
+```python
+>>> from cudf.core.buffer import Buffer
+>>> buf = Buffer(np.array([1, 2, 3]))
```

Comment:
This code shouldn't work anymore - we may only construct Buffer objects from int8 data. The intention was to prevent anyone from inadvertently passing `int64` when they think they're passing bytes.

Reply:
@brandon-b-miller can you add a suggestion here as to how we should change it?

Reply:
```suggestion
>>> buf = Buffer(np.array([1, 2, 3], dtype='uint8'))
```

Reply:
Nice catch! Fixed so that we view the inputs as `uint8` arrays

Reply:
👍